src/array/cpu/id_hash_map.h

```diff
@@ -0,0 +1,141 @@
+ * multi-threading to accelerate the insert and search speed. Currently it is
+ * only designed to be used in `ToBlockCpu` for optimizing.
+ *
+ * The hashmap should be used in two phases. With the first being creating the
+ * hashmap, and then init it with an id array. After that, searching any old ids
+ * to get the mappings according to your need.
+ *
+ * For example, for an array A with following entries:
+ * [98, 98, 100, 99, 97, 99, 101, 100, 102]
+ * Create the hashmap H with:
+ * `H = CpuIdHashMap()` (1)
+ * And Init it with:
+ * `U = H.Init(A)` (2)  (U is an id array used to store the unqiue
+ * ids in A).
+ * Then U should be (U is not exclusive as the element order is not
+ * guaranteed to be steady):
+ * [98, 100, 99, 97, 101, 102]
+ * And the hashmap should generate following mappings:
+ *  * [
+ *   {key: 98, value: 0},
+ *   {key: 100, value: 1},
+ *   {key: 99, value: 2},
+ *   {key: 97, value: 3},
+ *   {key: 101, value: 4},
+ *   {key: 102, value: 5}
+ * ]
+ * Search the hashmap with array I=[98, 99, 102]:
+ * R = H.Map(I) (3)
+ * R should be:
+ * [0, 2, 5]
+ **/
+template <typename IdType>
+class IdHashMap {
+ public:
+  /**
+   * @brief An entry in the hashtable.
+   */
+  struct Mapping {
+    /**
+     * @brief The ID of the item inserted.
+     */
+    IdType key;
+    /**
+     * @brief The value of the item inserted.
+     */
+    IdType value;
+  };
+
+  /**
+   * @brief Cross platform CAS operation.
+   * It is an atomic operation that compares the contents of a memory
+   * location with a given value and, only if they are the same, modifies
+   * the contents of that memory location to a new given value.
+   *
+   * @param ptr The pointer to the object to test and modify .
+   * @param old_val The value expected to be found in `ptr`.
+   * @param new_val The value to store in `ptr` if it is as expected.
+   *
+   * @return Old value pointed by the `ptr`.
+   */
+  static IdType CompareAndSwap(IdType* ptr, IdType old_val, IdType new_val);
+
+  IdHashMap();
+
+  IdHashMap(const IdHashMap& other) = delete;
+  IdHashMap& operator=(const IdHashMap& other) = delete;
+
+  /**
+   * @brief Init the hashmap with an array of ids.
+   * Firstly allocating the memeory and init the entire space with empty key.
+   * And then insert the items in `ids` concurrently to generate the
+   * mappings, in passing returning the unique ids in `ids`.
+   *
+   * @param ids The array of ids to be inserted as keys.
+   *
+   * @return Unique ids for the input `ids`.
+   */
+  IdArray Init(const IdArray ids);
+
+  /**
+   * @brief Find the mappings of given keys.
+   *
+   * @param ids The keys to map for.
+   *
+   * @return Mapping results corresponding to `ids`.
+   */
+  IdArray MapIds(const IdArray ids) const;
+
+ private:
+  void Next(IdType* pos, IdType* delta) const;
+
+  IdType MapId(const IdType id) const;
+
+  void Insert(IdType id, std::vector<int16_t>* valid, size_t index);
```

Comment:
Need to explain what is `valid`.  Also why a pointer to vector rather than `int16_t *`?  I don't think the size of the vector will change or anything.

Reply:
Add note and use NdArray(The elements is bool, don't use naive pointer to avoid memory operation) instead.